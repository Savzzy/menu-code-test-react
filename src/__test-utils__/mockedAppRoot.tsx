import { render, RenderResult } from "@testing-library/react";
import React from "react";
import { Provider } from "react-redux";
import { createStore } from "redux";
import { ThemeProvider } from "styled-components";
import globalTheme from "../globalTheme";
import rootReducer from "../reducers";
import { intialStore } from "../store";
import { Store } from "../types";

interface MockedAppRootOptions {
  mockStore: Store;
}

const mockedAppRoot = (
  ComponentTree: JSX.Element,
  options?: MockedAppRootOptions,
): RenderResult => {
  const AppContainer: React.FC = ({ children }) => (
    <Provider
      store={createStore(rootReducer, options?.mockStore || intialStore)}
    >
      <ThemeProvider theme={globalTheme}>{children}</ThemeProvider>
    </Provider>
  );

  return render(ComponentTree, { wrapper: AppContainer });
};

export default mockedAppRoot;
