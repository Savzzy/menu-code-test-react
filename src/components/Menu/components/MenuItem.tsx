import React, { useState, useEffect } from "react";
import styled from "styled-components";

import unsplash from "../../../api/unsplash";

const MenuItemContainer = styled.div`
  border-width: 1px;
  border-color: ${(props) => props.theme.colors.itemCardBorder};
  border-style: solid;
  display: flex;
  height: 90px;
  box-shadow: 2px 2px 4px 0px rgba(0, 0, 0, 0.1);
  border-radius: 5px;
  margin: 5px 0;
  cursor: pointer;
`;

const ImageContainer = styled.div`
  background: url("https://img.taste.com.au/_YKHzE3u/taste/2016/11/pumpkin-and-chive-soup-75984-1.jpeg");
  height: 90px;
  width: 100px;
  background-size: cover;
  background-position: center;
  background-repeat: no-repeat;
  border-radius: 5px 0 0 5px;
`;

const ItemDetailsContainer = styled.div`
  display: flex;
  flex-direction: column;
  padding: 10px;
`;

const ItemName = styled.h2`
  color: ${(props) => props.theme.colors.primaryText};
  font-size: 18px;
  font-weight: 700;
  margin: 0;
  font-weight: 400;
  flex-grow: 1;
`;

const ItemPrice = styled.h3`
  color: ${(props) => props.theme.colors.primaryText};
  font-size: 16px;
  margin: 0;
  font-weight: 400;
`;

interface MenuItemProps {
  itemId: number;
  itemName: string;
  itemPrice: number;
  priceUnit: string;
  onClick: (
    menuItemId: number,
    menuItemName: string,
    menuItemPrice: number,
  ) => void;
  addedToOrder?: boolean;
}

const MenuItem: React.FC<MenuItemProps> = ({
  itemId,
  itemName,
  itemPrice,
  priceUnit,
  onClick,
}) => {
  const [menuItemImage, setMenuItemImage] = useState(null);

  useEffect(() => {
    async function fetchMenuItem() {
      try {
        const response = await unsplash.get("search/photos", {
          params: { query: "cats" },
        });
      } catch (error) {
        console.log("unsplash API error");
      }
    }

    fetchMenuItem();
  });

  return (
    <MenuItemContainer
      onClick={() => {
        onClick(itemId, itemName, itemPrice);
      }}
    >
      <ImageContainer />
      <ItemDetailsContainer>
        <ItemName>{itemName}</ItemName>
        <ItemPrice>{`${priceUnit} ${itemPrice.toFixed(2)}`}</ItemPrice>
      </ItemDetailsContainer>
    </MenuItemContainer>
  );
};

export default MenuItem;
