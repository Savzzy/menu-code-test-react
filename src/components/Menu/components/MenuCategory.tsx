import equal from "fast-deep-equal";
import React, { useEffect, useRef } from "react";
import { Col, Grid, Row } from "react-flexbox-grid";
import { useDispatch } from "react-redux";
import styled from "styled-components";
import { addItemToOrder } from "../../../actions";
import MenuItem from "./MenuItem";

const CategoryName = styled.h2`
  color: ${(props) => props.theme.colors.primaryText};
  font-size: 20px;
  margin: 25px 0 20px 20px;
  text-transform: capitalize;
  font-weight: 400;
`;

const MenuItemsContainer = styled(Grid)``;

export interface MenuItemType {
  id: number;
  name: string;
  price: number;
}

interface MenuCategoryProps {
  menuCategory: string;
  menuItems: Array<MenuItemType>;
  registerOffset?: (menuCategory: string, offsetY: number) => void;
}

const MenuCategory: React.FC<MenuCategoryProps> = ({
  menuCategory,
  menuItems,
  registerOffset,
}) => {
  const categoryName = useRef(null);

  useEffect(() => {
    if (categoryName?.current && registerOffset) {
      registerOffset(
        menuCategory,
        categoryName.current.getBoundingClientRect().y,
      );
    }
  });

  const dispatch = useDispatch();

  return (
    <React.Fragment>
      <CategoryName ref={categoryName}>{menuCategory}</CategoryName>
      <MenuItemsContainer fluid>
        <Row>
          {menuItems.map((menuItem: MenuItemType) => {
            return (
              <Col xs={12} sm={6} md={4} key={menuItem.id}>
                <MenuItem
                  key={menuItem.id}
                  menuItem={menuItem}
                  priceUnit="Â£"
                  onClick={(menuItem: MenuItemType) => {
                    dispatch(addItemToOrder(menuItem, menuCategory));
                  }}
                />
              </Col>
            );
          })}
        </Row>
      </MenuItemsContainer>
    </React.Fragment>
  );
};

export default React.memo(MenuCategory, (prevProps, nextProps) => {
  return !equal(prevProps, nextProps);
});
