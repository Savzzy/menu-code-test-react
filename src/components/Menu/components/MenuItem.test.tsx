import React from "react";
import mockedAppRoot from "../../../__test-utils__/mockedAppRoot";
import MenuItem from "./MenuItem";
import { mockMenuItems } from "./MenuCategory.fixtures";
import { fireEvent } from "@testing-library/react";

describe("MenuItem", () => {
  it("renders correctly", () => {
    const { container } = mockedAppRoot(
      <MenuItem menuItem={mockMenuItems[0]} priceUnit="£" />,
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it("renderes passed menuItem name correctly", () => {
    const { getByText } = mockedAppRoot(
      <MenuItem menuItem={mockMenuItems[0]} priceUnit="£" />,
    );

    expect(getByText(mockMenuItems[0].name)).toBeInTheDocument();
  });

  it("test on menuItem Click", () => {
    const myMock = jest.fn();

    const { getByText } = mockedAppRoot(
      <MenuItem menuItem={mockMenuItems[0]} priceUnit="£" onClick={myMock} />,
    );

    fireEvent.click(getByText(mockMenuItems[0].name));

    expect(myMock.mock.calls.length).toBe(1);
    expect(myMock.mock.calls[0][0]).toBe(mockMenuItems[0]);

    myMock.mockClear();
  });

  it("renders delete icon when ordered props is true", () => {
    const { getByTestId } = mockedAppRoot(
      <MenuItem ordered={true} menuItem={mockMenuItems[0]} priceUnit="£" />,
    );
    expect(getByTestId("delete-icon")).toBeInTheDocument();
  });

  it("deletes the menuItem when clicked on delete icon", () => {
    const myMock = jest.fn();
    const { getByTestId } = mockedAppRoot(
      <MenuItem
        ordered={true}
        menuItem={mockMenuItems[0]}
        priceUnit="£"
        onDeleteIconClick={myMock}
      />,
    );

    fireEvent.click(getByTestId("icon-container"));
    expect(myMock.mock.calls.length).toBe(1);
    expect(myMock.mock.calls[0][0]).toBe(mockMenuItems[0].id);

    myMock.mockClear();
  });
});
