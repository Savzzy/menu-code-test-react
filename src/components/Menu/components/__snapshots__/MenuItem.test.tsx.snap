// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MenuItem renders correctly 1`] = `
<div
  class="sc-AxirZ eQsxdA"
  data-testid="menu_item"
>
  <div
    class="sc-AxiKw gtDHmr"
  />
  <div
    class="sc-AxhCb kqBmUn"
  >
    <h2
      class="sc-AxhUy jDDQbp"
    >
      Soup
    </h2>
    <h3
      class="sc-AxgMl dKATfS"
    >
      Â£ 3.00
    </h3>
  </div>
</div>
`;
